From: IrfanMohammad <irfan@graphiant.com>
Date: Fri, 24 Jan 2025 10:21:11 +0000
Subject: support delete of multi-key lists without key

sr_delete_item should be able to delete pushed operational data even
without having all the keys specified.

In this case, it should simply delete all entries that would normally
match.
---
 src/ly_wrap.c     |  2 +-
 src/sysrepo.c     | 57 +++++++++++++++++++++++++------------------------------
 tests/test_edit.c | 23 ++++++++++++++++++++++
 3 files changed, 50 insertions(+), 32 deletions(-)

diff --git a/src/ly_wrap.c b/src/ly_wrap.c
index bf2298d..676fde6 100644
--- a/src/ly_wrap.c
+++ b/src/ly_wrap.c
@@ -401,7 +401,7 @@ sr_lys_find_xpath(const struct ly_ctx *ctx, const char *xpath, uint32_t options,
         *valid = 1;
     }
 
-    if (lys_find_xpath(ctx, NULL, xpath, options, set)) {
+    if (lys_find_xpath(ctx, NULL, xpath, options, set) || !(*set) || !(*set)->count) {
         if (valid) {
             *valid = 0;
         } else {
diff --git a/src/sysrepo.c b/src/sysrepo.c
index 1ba8baa..c3d93f6 100644
--- a/src/sysrepo.c
+++ b/src/sysrepo.c
@@ -3575,7 +3575,7 @@ sr_delete_item(sr_session_ctx_t *session, const char *path, const sr_edit_option
     const char *op;
     const struct lysc_node *snode;
     struct lyd_node *node;
-    struct ly_set *set;
+    struct ly_set *set = NULL;
     uint32_t temp_lo = 0;
     uint32_t i;
     int rc;
@@ -3611,40 +3611,34 @@ sr_delete_item(sr_session_ctx_t *session, const char *path, const sr_edit_option
     }
 
     if (session->ds == SR_DS_OPERATIONAL) {
-        if ((err_info = sr_lys_find_path(session->conn->ly_ctx, path, NULL, &snode))) {
-            /* invalid path */
+        set = NULL;
+        /* check that the xpath is valid */
+        if ((err_info = sr_lys_find_xpath(session->conn->ly_ctx, path, LYS_FIND_NO_MATCH_ERROR, NULL, &set))) {
             goto cleanup;
         }
-        if ((snode->nodetype & (LYS_LIST | LYS_LEAFLIST)) && (path[strlen(path) - 1] != ']')) {
-            /* purge all the (leaf-)list instances */
-            set = NULL;
-            if (session->dt[session->ds].edit->tree &&
-                    (err_info = sr_lyd_find_xpath(session->dt[session->ds].edit->tree, path, &set))) {
-                goto cleanup;
-            }
-            if (set && set->count) {
-                for (i = 0; i < set->count; ++i) {
-                    sr_lyd_free_tree_safe(set->dnodes[i], &session->dt[session->ds].edit->tree);
-                }
-            } else if (opts & SR_EDIT_STRICT) {
-                /* not found */
-                sr_errinfo_new(&err_info, SR_ERR_NOT_FOUND, "No nodes \"%s\" found in session push oper data.", path);
-            }
-            ly_set_free(set, NULL);
-        } else {
-            /* just delete the selected node */
-            node = NULL;
-            if (session->dt[session->ds].edit->tree &&
-                    (err_info = sr_lyd_find_path(session->dt[session->ds].edit->tree, path, 0, &node))) {
-                goto cleanup;
-            }
-            if (node) {
-                sr_lyd_free_tree_safe(node, &session->dt[session->ds].edit->tree);
-            } else if (opts & SR_EDIT_STRICT) {
-                /* not found */
-                sr_errinfo_new(&err_info, SR_ERR_NOT_FOUND, "Node \"%s\" not found in session push oper data.", path);
+        if (!set || !set->count) {
+            /* due to a libyang bug, LYS_FIND_NO_MATCH_ERROR doesn't seem to work correctly */
+            sr_errinfo_new(&err_info, SR_ERR_NOT_FOUND, "No nodes \"%s\" found in session push oper data.", path);
+            goto cleanup;
+        }
+
+        ly_set_free(set, NULL);
+        set = NULL;
+
+        if (session->dt[session->ds].edit->tree &&
+                (err_info = sr_lyd_find_xpath(session->dt[session->ds].edit->tree, path, &set))) {
+            goto cleanup;
+        }
+        if (set && set->count) {
+            for (i = 0; i < set->count; ++i) {
+                sr_lyd_free_tree_safe(set->dnodes[i], &session->dt[session->ds].edit->tree);
             }
+        } else if (opts & SR_EDIT_STRICT) {
+            /* not found */
+            sr_errinfo_new(&err_info, SR_ERR_NOT_FOUND, "No nodes \"%s\" found in session push oper data.", path);
         }
+        ly_set_free(set, NULL);
+        set = NULL;
         goto cleanup;
     }
 
@@ -3666,6 +3660,7 @@ sr_delete_item(sr_session_ctx_t *session, const char *path, const sr_edit_option
             NULL, opts & SR_EDIT_ISOLATE);
 
 cleanup:
+    ly_set_free(set, NULL);
     if (session->dt[session->ds].edit && !session->dt[session->ds].edit->tree) {
         sr_release_data(session->dt[session->ds].edit);
         session->dt[session->ds].edit = NULL;
diff --git a/tests/test_edit.c b/tests/test_edit.c
index ab8bdde..d74c334 100644
--- a/tests/test_edit.c
+++ b/tests/test_edit.c
@@ -869,6 +869,29 @@ test_purge(void **state)
     assert_int_equal(ret, SR_ERR_OK);
     ret = sr_delete_item(st->sess, "/test:ll1", 0);
     assert_int_equal(ret, SR_ERR_OK);
+
+    ret = sr_set_item_str(st->sess, "/mod:container/list-enh[label='0'][date-and-time='1970-01-01T00:00:00Z']", NULL, NULL, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    ret = sr_apply_changes(st->sess, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    /* typo in xpath, should be caught */
+    ret = sr_delete_item(st->sess, "/mod:container/list-enha[label='0']", 0);
+    assert_int_equal(ret, SR_ERR_NOT_FOUND);
+
+    /* one key given in multi-key list, should be deleted */
+    ret = sr_delete_item(st->sess, "/mod:container/list-enh[label='0']", 0);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    ret = sr_apply_changes(st->sess, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    /* test that the data was actually deleted */
+    ret = sr_get_subtree(st->sess, "/mod:container/list-enh[label='0']", 0, &subtree);
+    assert_int_equal(ret, SR_ERR_OK);
+    assert_null(subtree);
+
     sr_session_switch_ds(st->sess, SR_DS_RUNNING);
 }
 
